File: scripts/ingest.py
#!/usr/bin/env python3
"""uploads/*.md 를 청크·임베딩해 documents 테이블에 적재 (Node 없이 실행 가능).
실행: python3 scripts/ingest.py   (.env 의 OPENAI_API_KEY, SUPABASE_* 필요)
표준 라이브러리만 사용."""
import json
import os
import sys
import urllib.request
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
UPLOADS = ROOT / "public" / "uploads"
EMBED_MODEL = "text-embedding-3-small"
CHUNK_SIZE = 800
CHUNK_OVERLAP = 150


def load_env():
    env = ROOT / ".env"
    if env.exists():
        for line in env.read_text(encoding="utf-8").splitlines():
            line = line.strip()
            if line and not line.startswith("#") and "=" in line:
                k, v = line.split("=", 1)
                os.environ.setdefault(k.strip(), v.strip())


def chunk_text(text):
    clean = text.replace("\r\n", "\n").strip()
    chunks, start = [], 0
    while start < len(clean):
        end = min(start + CHUNK_SIZE, len(clean))
        if end < len(clean):
            nl = clean.rfind("\n", start, end)
            if nl > start + CHUNK_SIZE * 0.5:
                end = nl
        piece = clean[start:end].strip()
        if piece:
            chunks.append(piece)
        if end >= len(clean):
            break
        start = max(0, end - CHUNK_OVERLAP)
    return chunks


def embed(text, api_key):
    req = urllib.request.Request(
        "https://api.openai.com/v1/embeddings",
        data=json.dumps({"model": EMBED_MODEL, "input": text}).encode("utf-8"),
        headers={"Content-Type": "application/json", "Authorization": f"Bearer {api_key}"},
        method="POST",
    )
    with urllib.request.urlopen(req, timeout=60) as resp:
        return json.loads(resp.read())["data"][0]["embedding"]


def supabase_request(method, path, body=None, prefer=None):
    url = os.environ["SUPABASE_URL"].rstrip("/") + path
    key = os.environ["SUPABASE_SERVICE_ROLE_KEY"]
    headers = {
        "apikey": key,
        "Authorization": f"Bearer {key}",
        "Content-Type": "application/json",
    }
    if prefer:
        headers["Prefer"] = prefer
    data = json.dumps(body).encode("utf-8") if body is not None else None
    req = urllib.request.Request(url, data=data, headers=headers, method=method)
    with urllib.request.urlopen(req, timeout=60) as resp:
        raw = resp.read()
        return json.loads(raw) if raw else None


def main():
    load_env()
    if not os.environ.get("OPENAI_API_KEY"):
        sys.exit("❌ OPENAI_API_KEY 가 .env 에 없습니다.")
    if not os.environ.get("SUPABASE_URL") or not os.environ.get("SUPABASE_SERVICE_ROLE_KEY"):
        sys.exit("❌ SUPABASE_URL / SUPABASE_SERVICE_ROLE_KEY 가 .env 에 없습니다.")
    api_key = os.environ["OPENAI_API_KEY"]

    print("기존 documents 삭제 중...")
    # id != 0 인 전체 삭제
    supabase_request("DELETE", "/rest/v1/documents?id=gt.0", prefer="return=minimal")

    total = 0
    for f in sorted(UPLOADS.glob("*.md")):
        chunks = chunk_text(f.read_text(encoding="utf-8"))
        print(f"📄 {f.name} → {len(chunks)} 청크")
        for content in chunks:
            embedding = embed(content, api_key)
            supabase_request(
                "POST", "/rest/v1/documents",
                body={"source": f.name, "content": content, "embedding": embedding},
                prefer="return=minimal",
            )
            total += 1
    print(f"✅ 완료: {total} 청크 적재됨")


if __name__ == "__main__":
    main()

File: scripts/ingest.js
// uploads/*.md 를 청크·임베딩해 documents 테이블에 적재한다.
// 실행: node scripts/ingest.js   (.env 의 OPENAI_API_KEY, SUPABASE_* 필요)
require('dotenv').config();
const fs = require('fs');
const path = require('path');
const OpenAI = require('openai');
const { getSupabase } = require('../lib/supabase');
const { EMBED_MODEL } = require('../lib/rag');

const UPLOADS = path.join(__dirname, '..', 'public', 'uploads');
const CHUNK_SIZE = 800; // 문자 기준
const CHUNK_OVERLAP = 150;

// 문단 경계를 최대한 보존하며 청크로 분할
function chunkText(text) {
  const clean = text.replace(/\r\n/g, '\n').trim();
  const chunks = [];
  let start = 0;
  while (start < clean.length) {
    let end = Math.min(start + CHUNK_SIZE, clean.length);
    // 단어/문단 중간이 잘리지 않게 가까운 줄바꿈에서 끊기
    if (end < clean.length) {
      const nl = clean.lastIndexOf('\n', end);
      if (nl > start + CHUNK_SIZE * 0.5) end = nl;
    }
    const piece = clean.slice(start, end).trim();
    if (piece) chunks.push(piece);
    if (end >= clean.length) break;
    start = end - CHUNK_OVERLAP;
    if (start < 0) start = 0;
  }
  return chunks;
}

async function main() {
  const supabase = getSupabase();
  if (!supabase) {
    console.error('❌ SUPABASE_URL / SUPABASE_SERVICE_ROLE_KEY 가 .env 에 없습니다.');
    process.exit(1);
  }
  if (!process.env.OPENAI_API_KEY) {
    console.error('❌ OPENAI_API_KEY 가 .env 에 없습니다.');
    process.exit(1);
  }
  const client = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });

  // 기존 documents 비우고 새로 적재 (재실행 시 중복 방지)
  console.log('기존 documents 삭제 중...');
  await supabase.from('documents').delete().neq('id', 0);

  const files = fs.readdirSync(UPLOADS).filter((f) => f.toLowerCase().endsWith('.md'));
  let total = 0;

  for (const file of files) {
    const text = fs.readFileSync(path.join(UPLOADS, file), 'utf-8');
    const chunks = chunkText(text);
    console.log(`📄 ${file} → ${chunks.length} 청크`);

    for (const content of chunks) {
      const emb = await client.embeddings.create({ model: EMBED_MODEL, input: content });
      const embedding = emb.data[0].embedding;
      const { error } = await supabase.from('documents').insert({ source: file, content, embedding });
      if (error) {
        console.error(`  ⚠️ 적재 실패: ${error.message}`);
      } else {
        total++;
      }
    }
  }

  console.log(`✅ 완료: ${total} 청크 적재됨`);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
